### ISSUE TITLE
Caching seems to be project-independent

### ISSUE THREAD

**Issue by ohthehugemanatee** [ericpaulbishop/redmine_git_hosting#43]
When I set any values in the new caching fields, the Repository tab will always show the repo for the most recent commit, no matter which project that commit was on.  For example, if I commit to project A, then visit project B's Repository tab, I will see content from project A.  I see that project A repo when visiting projects C, D, and Q also.

Secondarily, I suspect the cache is not clearing on commit.  I can commit changes which I can see in the repository, but which do not appear in Redmine's repository tab.  I've tried forcing it to re-read the commit log with ruby script/runner "Repository.fetch_changesets" -e production , and through the web services command http://projects.example.com/sys/fetch_changesets?key=<123456ABCDEFG>&id=project .  This could be related - ie I could just be seeing cache when I look at the repository tab.  I tried setting the caching values on the settings page to 0 and to 1, but this behavior is unchanged.  These are small repos, so it's totally possible that the cache fits into 1MB. :|  A "do not cache" selection option would be helpful here.

I did have a problem after the yourchili install - trying to commit to repos gave an error that post-commit couldn't find update.secondary.  So I added the script in ~git/.gitolite/hooks/common as an empty, executable file, and ran gl-setup as the git user.  I didn't expect it to cause any problems, but if yourchili does anything fancy with the hooks that might be related.

**Comment by ohthehugemanatee** [comment]
ah - just noticed that I could set caching to 1 minute.  That does indeed seem to solve the problem.  This is probably a very simple issue in the cache table lookup.  It might just need a where clause for project ID.   

Though now, I see the repo file listing with all the commit information... but the block at the bottom which lists commits does not appear.  I can see the commits if I hit the link to "view all revisions", so I'm not clear what the difference is between the SQL statements that generate those two displays of the same information.

**Comment by ericpaulbishop** [comment]
There is already a proj_identifier column in the git_caches table, and lookups are already done on a per-project basis.  It does a lookup based on the full command which includes the repo path, but it clears all cached data for a project when the post-receive hook is called and this is done based on the proj_identifier.

I've tried to reproduce this error you describe, but this is another error I can't reproduce... it works fine for me.

**Comment by ohthehugemanatee** [comment]
well one biggie would be that I can't find your post-receive hook.  Looking in /srv/git/repositories/projectname/hooks , it appears that when I updated the hooks using .gitolite/hooks/common/\* , it wiped out the hooks that yourchili scripted on a per-project basis.  

This can be easily avoided by putting a copy of the per-project hooks into ~git/.gitolite/hooks/common, so people don't overwrite the redmine hooks unwittingly like that.   But then of course it assumes that redmine is handling ALL your repos.  That's fine for me, but not for everyone.  Maybe we could alter the hook so it tests to make sure it's a redmine project first?  

Meanwhile, I added 2 lines to the top of my ~git/.gitolite/hooks/common/post-update script (which is where I update my dev checkout of Master branch anyway):

```
sudo -u www-data ruby "/srv/www/projects.example.com/chili/script/runner" -e production "GitHosting::run_post_receive_hook(\"$GL_REPO\")" >/dev/null 2>&1
/usr/bin/wget -O - -q -t 1 "http://projects.example.com/sys/fetch_changesets?key=123456ABCDEF"
```

This fixed my problem - I can use repo caching now, and the little repo view at the bottom of the Repositories tab works again.  This brings me to 3 questions:

1) Clearly the problem with the little repo view is connected to caching. Does the caching code patch the Repository tab at all? What happens if caching is borked, as mine was?

2) do I even need the fetch_changesets line in this script?  It runs very quickly, but I'm not clear if it's duplicating some of the effort of your post_receive_hook action.

3) Is it normal for the post_receive_hook action to be slow?  It takes several seconds for me. By my reading that's what https://github.com/ericpaulbishop/redmine_git_hosting/pull/37 is about.  It might be worthwhile to look into firing this asynchronously, so it doesn't slow down the commit process. 

I apologize for all the questions lately.  I'm just a sysadmin who usually works in PHP, Redmine/Chili is my only exposure to Ruby so I'm having a hard time troubleshooting for myself.  I appreciate all the support and effort you've put in, and I'm trying to give enough info to be helpful. :)

**Comment by s0undt3ch** [comment]
This is a good example on which my hooks aproach handles better :)

It installs the needed hooks into gitolite's 'hooks/common', and although the update is not made async, ie, it's still sync, it's way much faster and if python is available to the git user, it's even a bit more faster.

So, Eric, how about it? :)

**Comment by ericpaulbishop** [comment]
A new approach that uses a different, (optionally asynchronous) method for running hooks has been implemented in v0.4.0.  This should solve these issues.
